astropy/utils/collections.py

```diff
@@ -56,3 +69,101 @@ def extend(self, x):
         for item in x:
             self._assert(item)
             super().append(item)
+
+
+def _locked_mutator(instance_method):
+    @wraps(instance_method)
+    def inner(self, *args: Any, _owned: bool = False, **kwargs):
+        if not _owned:
+            cls_name = self._owner_class.__name__
+            msg = (
+                f"Directly mutating {cls_name}.{self._attr_name} "
+                f"via {instance_method.__name__} is deprecated and will raise an "
+                "error in the future. "
+            )
+
+            if (repl := self._replacements.get(self._attr_name)) is not None:
+                msg += f"Please use {cls_name}.{repl} instead"
+            else:
+                msg += (
+                    "There is currently no planned replacement for this method. "
+                    "If you need one, please open a feature request at "
+                    "https://github.com/astropy/astropy/issues/new/choose"
+                )
+
+            warnings.warn(
+                msg,
+                category=DeprecationWarning,
+                stacklevel=2,
+            )
+        return instance_method(self, *args, **kwargs)
+
+    return inner
+
+
+class _OwnedList(list):
+    def __init__(
+        self,
+        *args: Any,
+        owner_class: type,
+        public_attr_name: str,
+        replacements: dict[str, str],
+        **kwargs,
+    ):
+        # storing metadata to make warnings and errors more useful
+        self._owner_class = owner_class
+        self._attr_name = public_attr_name
+        self._replacements = replacements
+        super().__init__(*args, **kwargs)
+
+    @_locked_mutator
+    def append(self, object, /):
+        return super().append(object)
+
+    @_locked_mutator
+    def clear(self, /):
+        return super().clear()
+
+    @_locked_mutator
+    def extend(self, iterable, /):
+        return super().extend(iterable)
+
+    @_locked_mutator
+    def insert(self, index, object, /):
+        return super().insert(index, object)
+
+    @_locked_mutator
+    def pop(self, index=-1, /):
+        return super().pop(index)
+
+    @_locked_mutator
+    def remove(self, value, /):
+        return super().remove(value)
+
+    @_locked_mutator
+    def reverse(self, /):
+        return super().reverse()
+
+    @_locked_mutator
+    def sort(self, /, *, key=None, reverse=False):
+        return super().sort(key=key, reverse=reverse)
+
+    @_locked_mutator
+    def __iadd__(self, value: Iterable) -> Self:
+        return super().__iadd__(value)
+
+    @_locked_mutator
+    def __imul__(self, value: SupportsIndex) -> Self:
+        return super().__imul__(value)
+
+    @_locked_mutator
+    def __setitem__(self, key, value, /) -> Self:
+        return super().__setitem__(key, value)
+
+    @_locked_mutator
+    def __delitem__(self, key: SupportsIndex | slice) -> None:
```

Comment:
A note on type annotations: I didn't intend to add them, but got them for free from my editor's supercharged auto complete.